fe/fe-core/src/main/java/com/starrocks/scheduler/mv/IMTCreator.java

```diff
@@ -0,0 +1,231 @@
+            IMTCreatorVisitor visitor = new IMTCreatorVisitor();
+            visitor.stmt = stmt;
+            visitor.view = view;
+            visitor.visit(visitor, optExpr);
+            return visitor.result;
+        }
+
+        private void visit(IMTCreatorVisitor visitor, OptExpression optExpr) {
+            for (OptExpression child : optExpr.getInputs()) {
+                visitor.visit(child, null);
+            }
+            visitor.visit(optExpr, null);
+        }
+
+        @Override
+        public Void visit(OptExpression optExpr, Void ctx) {
+            return null;
+        }
+
+        @Override
+        public Void visitPhysicalStreamJoin(OptExpression optExpr, Void ctx) {
+            // Join does not need any IMT
+            return null;
+        }
+
+        /**
+         * StreamAgg needs two IMT
+         * 1. State Table: store the state of all aggregation functions for non-retractable ModifyOp
+         * 2. Detail Table: for retractable ModifyOp
+         * <p>
+         * TODO(murphy) create detail table
+         */
+        @Override
+        public Void visitPhysicalStreamAgg(OptExpression optExpr, Void ctx) {
+            MVOperatorProperty property = optExpr.getMvOperatorProperty();
+            ColumnRefFactory columnRefFactory = stmt.getColumnRefFactory();
+            Map<String, String> properties = view.getProperties();
+
+            if (!property.getModify().isInsertOnly()) {
+                throw UnsupportedException.unsupportedException("Not support retractable aggregate");
+            }
+
+            // Duplicate/Primary Key
+            KeyInference.KeyProperty bestKey = property.getKeySet().getBestKey();
+
+            // Columns
+            // TODO(murphy) create columns according to StreamAgg requirement
+            List<Column> keyColumns = new ArrayList<>();
+            List<ColumnDef> columnDefs = new ArrayList<>();
+            ColumnRefSet columnRefSet = optExpr.getOutputColumns();
+            for (int columnId : columnRefSet.getColumnIds()) {
+                ColumnRefOperator refOp = columnRefFactory.getColumnRef(columnId);
+
+                Column newColumn = new Column(refOp.getName(), refOp.getType());
+                boolean isKey = bestKey.columns.contains(refOp);
+                newColumn.setIsKey(isKey);
+                newColumn.setIsAllowNull(false);
+                if (isKey) {
+                    keyColumns.add(newColumn);
+                }
+
+                TypeDef typeDef = TypeDef.create(refOp.getType().getPrimitiveType());
+                ColumnDef columnDef = new ColumnDef(refOp.getName(), typeDef);
+                columnDef.setIsKey(isKey);
+                columnDef.setAllowNull(!isKey);
+                columnDefs.add(columnDef);
+            }
+
+            // Key type
+            KeysType keyType = property.getModify().isInsertOnly() ? KeysType.DUP_KEYS : KeysType.PRIMARY_KEYS;
```

Comment:
How can StreamAgg IMTState table's key type be DUP_KEYS?  it must be PRIMARY_KEYS?

Reply:
fixed